### ISSUE TITLE
Improve modularity by creating architecure guidelines

### ISSUE THREAD

**Issue by mossmaurice** [eclipse-iceoryx/iceoryx#590]
## Brief feature description

iceoryx is developed in several CMake packages that aim to be independent. Current state is there are cross dependencies between each of them and no architecture guidelines have been created so far.

## Detailed information

Analysed iceoryx repository with [Lattix Architect](https://www.lattix.com/products-architecture-issues/). [Attached](https://github.com/eclipse-iceoryx/iceoryx/files/6071770/iceoryx_b929aa_Architecture.Metrics_Metrics.Architecture.xlsx) are the the dependecy structure matrix and the architecture metrixs of Lattix (commit b929aa8d5022471a318cdc8e6994f1cdd6aa3850).

### General

* [x] Analyse Lattix
  * [x] Dependecy structure matrix (DSM)
  * [x] Architectual metrics
  * [x] Conceptual architecture
* [x] Namespaces & CMake package/library refactoring
* [x] Add CMake target for quality level 5/playground features and move "maybe obsolote" constructs there
  * [x] Add script to compare used namespaces to a whitelist in binary (`nm -C myBinary -f posix` with C++filt for demangling) to make sure no experimental/playground code is baked into releases
* [ ] RouDi modularisation #539
* [x] Create architecture documentation
  * [x] PlantUML or plain ASCII art? 
  * [x] Package view
  * [x] Layer view
* [ ] Summarise guidelines and write a CI checker
  * [ ] Remove `iox::helper` namespace
* [ ] Create `iceoryx_posh_core` CMake target and address the `@todo #590` in the CMake files
* [ ] Remove unused files and move quality level 2 files to a separate CMake package

### Lattix Architect findings

* [x] ~~Move POSH specific errors from iceoryx_utils to iceoryx_posh~~ Moved to #1099 
* [x] `roudi_gtest.hpp` includes `iceoryx_posh/internal/roudi_environment/roudi_environment.hpp`
* [ ] Move `RouDiEnvironment` to test folder & namespace
* [ ] `iceoryx_hoofs` does not have a clear architecture and shall be better structured

**Comment by mossmaurice** [comment]
FYI @dlo-86 
[Structure101](https://structure101.com) would be another option besides Lattix to analyse the architecture. However, I've had a detailed look at Lattix and found the DSM and its metrics quite helpful.

**Comment by elBoberido** [comment]
@mossmaurice do you think this is related to iceoryx_hoofs?

**Comment by mossmaurice** [comment]
> do you think this is related to iceoryx_hoofs?
@elBoberido 

Nope, I think this issue is mainly `iceoryx_posh` related. As discussed today, any architecture guidelines and their respective CI checkers should be implemented for a complete basic example like `icedelivery` and not just for `iceoryx_hoofs`. The only thing I can think about is to split up the `iceoryx_hoofs` modules / namspaces into several CMake targets. But that should probably go into a new issue. Shall I create one?

**Comment by mossmaurice** [comment]
> yes, please create the issue.

[Done](https://github.com/eclipse-iceoryx/iceoryx/issues/1391) :heavy_check_mark:
